### PR TITLE
Add GetAll methods for RBAC

### PR DESCRIPTION
**Pull request by knanao**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/model/project.go [hunk 1]
```diff
@@ -395,3 +395,57 @@ func ValidateUserGroups(groups []*ProjectUserGroup) error {
...
+		}
+		v = append(v, group)
+	}
+	v = append(v, p.UserGroups...)
+
+	p.UserGroups = v
+	return p
+}
+
+func (p *Project) WithBuiltinRBACRoles() *Project {
```
Comment (nghialv) [inline pkg/model/project.go:434]: I feel that we should avoid updating the Project model like this.
Instead, we can have a function like `ListAllRBACRoles() []ProjectRBACRole` which merges and returns the list of all custom and built-in roles.

Comment (nghialv) [inline pkg/model/project.go:434]: I feel that we should avoid updating the Project model like this.
Instead, we can have a function like `ListAllRBACRoles() []ProjectRBACRole` which merges and returns the list of all custom and built-in roles.

File: pkg/model/project.go [hunk 2]
```diff
@@ -395,3 +395,57 @@ func ValidateUserGroups(groups []*ProjectUserGroup) error {
 	}
 	return nil
 }
+
+func (p *Project) MergeRBAC() *Project {
```
Comment (nghialv) [inline pkg/model/project.go:399]: Same as https://github.com/pipe-cd/pipecd/pull/3487/files#r843403939
we can have `ListAllUserGroups() []ProjectUserGroup`.
Wdyt?

Comment (nghialv) [inline pkg/model/project.go:399]: Same as https://github.com/pipe-cd/pipecd/pull/3487/files#r843403939
we can have `ListAllUserGroups() []ProjectUserGroup`.
Wdyt?

File: pkg/model/project.go (hunk 1 above)
Comment (knanao) [reply pkg/model/project.go:434]: Alright, Let's adopt that one.
but in this situation, Is `GetAllRBACRoles` better?

Comment (nghialv) [reply pkg/model/project.go:434]: That is fine to me as well. 👍

File: pkg/model/project.go (hunk 2 above)
Comment (knanao) [reply pkg/model/project.go]: Thank you for your suggestion. Addressed https://github.com/pipe-cd/pipecd/pull/3487/commits/ed623723d6ff26d6dd6a394714b48690b696c650

File: pkg/model/project.go [hunk 3]
```diff
@@ -395,3 +395,55 @@ func ValidateUserGroups(groups []*ProjectUserGroup) error {
...
 }
+
+func (p *Project) GetAllUserGroups() []*ProjectUserGroup {
+	rbac := p.Rbac
+	if rbac == nil {
+		return p.UserGroups
+	}
+
+	size := len(p.UserGroups) + 3 // add the count of rbac teams.
+	v := make([]*ProjectUserGroup, 0, size)
```
Comment (nghialv) [inline pkg/model/project.go:406]:
```suggestion
	// The full list also contains 3 legacy user groups.
	all := make([]*ProjectUserGroup, 0, len(p.UserGroups) + 3)
```

File: pkg/model/project.go [hunk 4]
```diff
@@ -395,3 +395,55 @@ func ValidateUserGroups(groups []*ProjectUserGroup) error {
...
+	if rbac.Viewer != "" {
+		group := &ProjectUserGroup{
+			SsoGroup: rbac.Viewer,
+			Role:     builtinRBACRoleViewer.String(),
+		}
+		v = append(v, group)
+	}
+	v = append(v, p.UserGroups...)
+
+	return v
```
Comment (nghialv) [inline pkg/model/project.go:430]:
```suggestion
	if rbac.Admin != "" {
		all = append(all, &ProjectUserGroup{
			SsoGroup: rbac.Admin,
			Role:     builtinRBACRoleAdmin.String(),
		})
	}
	if rbac.Editor != "" {
		all = append(all, &ProjectUserGroup{
			SsoGroup: rbac.Editor,
			Role:     builtinRBACRoleEditor.String(),
		})
	}
	if rbac.Viewer != "" {
		all = append(all, &ProjectUserGroup{
			SsoGroup: rbac.Viewer,
			Role:     builtinRBACRoleViewer.String(),
		})
	}
	all = append(all, p.UserGroups...)

	return all
```

File: pkg/model/project.go [hunk 5]
```diff
@@ -395,3 +395,55 @@ func ValidateUserGroups(groups []*ProjectUserGroup) error {
...
+
+	// Initialize the capcity since the length of RbacRoles will continue to increase.
+	size := len(p.RbacRoles) + len(builtin)
+	v := make([]*ProjectRBACRole, 0, size)
+	// Set built-in rbac role.
+	v = append(v, builtin...)
+	// Set custom rbac role.
+	v = append(v, p.RbacRoles...)
+
+	return v
```
Comment (nghialv) [inline pkg/model/project.go:448]:
```suggestion
	all := make([]*ProjectRBACRole, 0, len(p.RbacRoles) + len(builtin))
	// Set built-in rbac role.
	all = append(all, builtin...)
	// Set custom rbac role.
	all = append(all, p.RbacRoles...)

	return all
```

**Comment by nghialv** [comment]
@knanao Look nice. Just some nits. 💯

**Comment by nghialv** [comment]
Nice.
/lgtm

**Comment by khanhtc1202** [comment]
Looks neat
/approve

**Pull request merged**